### DISCUSSION TITLE
Nightscout on Google Cloud

### DISCUSSION THREAD

**Discussion by DaNitroo** [NightscoutFoundation/xDrip#2676] [General]
Hello.
First I would like to introduce myself and thank you. My name is Daniel and i`m from Germany and my son has had type 1 diabetes for 6 weeks. You have really created something great with your community! It is incredible how you help people with this and thanks to you great inventors and programmers you have really made our lives much safer! Thanks a lot for this!

I followed this great guide https://navid200.github.io/xDrip/docs/Nightscout/GoogleCloud.html. Everything (freedns, google cloud, VM, Nightscout install, Nightscout setting up) works great and my nightscout is now running in the google cloud. many many thanks for that!!
2 questions stayed with me despite the great instructions (sorry, but this is the first time in my life that I have to do with computers. I usually repair cars):

My first question: I backed up the MongoDB via point 6-2 and then downloaded it to my PC via the terminal. That worked fine. But I would be interested to know where I actually download this data or where is this data stored. In the google cloud? In the nightscout variable "database_size" i still have to specify the size of the database and i don't know where i can find it?

My second question: I have just created a MongoDB but I have not yet entered it in the variables at nightscout and now I see that my nightscout has been running for 2 days without MongoDB and the data seems to be saving all. Do I actually need the MongoDB?

**Comment by Navid200** [comment]
Hi,

When you follow the guide to set up Nightscout on Google Cloud, your database is set up for you through the installation process.
You don't need to sign up for MongoDB.

If you use Nightscout on Google Cloud for a while, and you need to delete it, you should backup the database so that you don't lose your data because the database is on the virtual machine.
So, that's the reason for this page: https://navid200.github.io/xDrip/docs/Nightscout/DatabaseBackup.html

If you have made such a backup and now have created a new machine and would like to restore such a backup into it, you use this page:
https://navid200.github.io/xDrip/docs/Nightscout/DatabaseRestore.html

If you have been using Nightscout on a different platform, that is still functional, and now have set up Nightscout on Google Cloud, and you want to transfer your database content to this setup, you can use this:
https://navid200.github.io/xDrip/docs/Nightscout/NS_Transfer.html

If you were using Heroku and now have set up Nightscout on Google Cloud and you want to transfer your database content to this, you may be able to use this: https://navid200.github.io/xDrip/docs/Nightscout/HerokuGone.html

Please let me know if I misunderstood your questions.

**Reply by DaNitroo** [reply]
Wow. Thank you very much for your fast and professional answer! :-)

You got my questions right.
So I'm happy, then I can save myself the work with MongoDb :-)

I'm just a little worried about the backup. We are currently using the Dexcom G7 and soon the Libre 3 and I was looking for a way to collect all the data, no matter where it came from, so that my son can also see it later and still be able to access it many years from now.

How big is my Google Cloud database that was automatically set up here and can I see somewhere how full it is already?

I was shown 30GB, that won't be right for 0€, right?

Thanks again for the help. I'm really very happy that there are people like you who make our lives easier!

**Reply by Navid200** [reply]
If you followed the guide you have linked to above, the disk space is 30GB.  However, the Linux operating system itself and the other packages take some of that space.

So, we set the Nightscout variable associated with the capacity available for the MongoDB database stored on the machine to 20GB.
If you open Nightscout in a browser, you can see what percentage of that capacity is full already.

If you open your status page, you can see how much of the hard drive space is occupied.  https://navid200.github.io/xDrip/docs/Nightscout/Status.html
In the image you can see on that page, it shows that 17% of the 30GB disk is occupied.  

I have diabetes.  The reason I use Nightscout is so that I can have a record of all my readings for now and the future.
Of course this means I have to protect the database.  That's why the backup page was one of the very first pages I added to our site.
You should never ever delete the virtual machine before backing up the database.

It is my dream to add a feature to add the ability to do automatic backups to Google Drive.  But, that is a long-term project.  I can make no promises if that will even ever be possible.

**Reply by DaNitroo** [reply]
Das ist ja wirklich super. Besser geht es ja nicht mehr (außer google drive sync, aber man kann doch auch nicht alles haben) :-)
Die Anzeige für den Festplattenspeicher und die Hinterlegung von 20GB hatte ich natürlich bemerkt, aber ich hatte es nicht für wahr gehalten, dass ich wirklich einen so großen Speicher kostenlos bekomme! :-) Danke nochmal, es funktioniert jetzt alles perfekt.

Schade ist nur, dass bei Dexcom Clarity alle Daten von Anfang an drin sind, aber bei Nightscout sind nur die letzten Tage drin seit dem wir XDRIP benutzten. Die XDRIP Autofill Funktion ändert daran nichts und bei Dexcom kann ich nur eine csv.Datei exportieren und die kann ich bei Nightscout nicht importieren, oder?

**Reply by Navid200** [reply]
Translated (to English) using Google:  
  
---  
  
That's really great. It doesn't get any better (apart from google drive sync, but you can't have everything) :-)
Of course, I had noticed the display for the disk space and the deposit of 20GB, but I had not believed that I really get such a large memory for free! :-) Thanks again, everything works perfectly now.

It's just a pity that Dexcom Clarity includes all data from the start, but Nightscout only includes the last few days since we started using XDRIP. The XDRIP Autofill function doesn't change that and with Dexcom I can only export a csv file and I can't import it with Nightscout, right?

**Reply by Navid200** [reply]
@tzachi-dar @psonnera 
How can one import CSV?
Can xDrip import CSV?
Can Nightscout import CSV?

**Reply by psonnera** [reply]
You can do that on Mongo databases with external programs like this
![image](https://user-images.githubusercontent.com/33809361/218865156-013b0c80-0b09-457f-917c-ab9229c67742.png)

**Reply by psonnera** [reply]
But you need to match the entries format in the CSV

**Reply by DaNitroo** [reply]
Thank you for your quick response.

Editing columns etc. in Excel would not be a problem.

I followed the instructions to install nightscout on a vm in Google Cloud, the MongoDB is already integrated there as I understand it. So I don't have access to that database, do I?

**Reply by psonnera** [reply]
NS in GC doesn't expose the database AFAIK. I don't know if you could use the backup generated and edit it...
Another (heavy) solution would be to create a temporary Railway Nightscout with a mongo database and import your data inside it, then copy it to your GC DB with the utility. Maybe I'm over-engineering it and hopefully there's more simple.

**Reply by Navid200** [reply]
I'm not sure what you mean by not exposing the database.
Nightscout on Google Cloud uses official MongoDB.
That's why we can import a database from Heroku as explained here:  https://navid200.github.io/xDrip/docs/Nightscout/HerokuGone.html

**Reply by psonnera** [reply]
@Navid200 I mean Atlas is exposed: the connection string is available for any computer connected to the internet with the connection string. Nightscout-VPS has a loca database that is isolated inside the VPS (and that's making it very secure and safe) but there's no public connection string allowing you to access it from outside the VPS, like with the software I showed above. If you want to have a public access to the database you'd need to create a new user (or use the default one https://github.com/jamorham/nightscout-vps/blob/16187a5ad9be020e19dc4c297056d188b5660582/NS_Install.sh#L44 but you can see nybody can know it) and open port 27017 to access it (and much probably more in nginx). Bottom line: a lot of modifications.

**Reply by Navid200** [reply]
I may have lost the track of the conversation.
I thought the objective was to transfer readings from Clarity to the database inside the Nightscout on Google Cloud.
I don't know why that would require to access the database in public.  If we can get the data out of Clarity, why can we not import it into the MongoDB database inside the virtual machine?

I don't know how to import a csv file into MongoDB.  If there is a way, we can import.  I am even happy to add a menu item if needed.  Isn't this how to do it?
https://www.mongodb.com/developer/products/mongodb/mongoimport-guide/

**Reply by DaNitroo** [reply]
That doesn't look bad, but I don't quite get it. The goal is this: I have all of my son's CGM data at Dexcom Clarity and I also can export a csv file from there. But I don't have this data in XDRIP and can therefore not export it to Nightscout using the backfill option. So if I find a way to export a Nightscout csv file from MongoDB, I can look at it, then bring my existing csv file with my Clarity data into the same format and then import it back into my MongoDB. But I can't find the MongoDB in the GC. Doing the shown way from "psonnera" via a detour and railway seems to be complex, but easy and understandable for a beginner like me. Many thanks for this tip!!

I don't care if it takes a long time, I just have to understand it and sometimes that's really difficult for me :-)

**Reply by Navid200** [reply]
I am going to export my xDrip data into a csv file and find a way to import it into Nightscout.
Then, I will let you know how.

**Reply by Navid200** [reply]
@DaNitroo Open a terminal and exit the menu.
In Linux, type mongo shell
Later on after you are done, you can type exit to exit mongo shell.

Now, in mongo shell, type show dbs
You will see this:
![Screenshot 2023-02-14 194819](https://user-images.githubusercontent.com/51497406/218896534-10042620-91c8-4e96-a2fa-389241b6ae5f.png)

You can see that we have called Nightscout just Nightscout.

now, type use Nightsocut.
Now, type show collections.  And that shows you this.
![Screenshot 2023-02-14 195027](https://user-images.githubusercontent.com/51497406/218896803-9c9bb7dc-b872-4b0a-ae2a-6a8c5bbc010e.png)

So, you should be able to access MongoDB on your machine.

**Reply by DaNitroo** [reply]
I can open a terminal and do exact what you tell me. When I enter "use nightscout" it shows "switched to db nightscout". What had I to do next to output an csv file?


It would be so great if I you can show me, how I can download a csv. Then I would change my Dexcom csv to the same formatting and upload it back to Nightscout. Then I would have all my data in there forever :-)

**Reply by DaNitroo** [reply]
I just saw that I had a translation error. but when I type show collections I don't see any. do they have to be created somewhere first?

**Reply by Navid200** [reply]
No, they are created by Nightscout.

The only difference between Nightscout on Google Cloud and Heroku or any other platform is that we have changed the AUTH_DEFAULT_ROLES variable from readable to denied.
You can change that back to readable if you want.  Then, your cost for the month may become $1.  You can change it back after you are done with your transfer.

It is a misunderstanding to think that there is something special about Nightscout on Google Cloud that makes it different other than that single variable, which you can change if you like.
I have gone into detail about that here: https://navid200.github.io/xDrip/docs/Nightscout/Auth_Default_Roles.html

Transferring data from Clarity to Nightscout is a question for the Nightscout team.  We are not the Nightscout team.  We are the xDrip team.  
I personally have no experience with Clarity and cannot help you.  That's why I tagged others.
But, the best place to ask this question would be in a Nightscout forum.  This is one example:
https://www.facebook.com/groups/cgminthecloud
And I strongly advise against focusing on Nightscout on Google Cloud because many can easily get confused and think that only we may know how to do it.

**Comment by DaNitroo** [comment]
ok, i understand that
Can you tell me where I can find the MongoDB URI in my Google Cloud VM?
Then I found the solution to my problems

<img width="449" alt="image" src="https://user-images.githubusercontent.com/122942359/219122100-7eb202c0-3fc4-4802-9828-ea8362b902de.png">

**Reply by Navid200** [reply]
We don't use MongoDB on Atlas.
We use MongoDB locally on the virtual machine.
There is no URL.

**Reply by DaNitroo** [reply]
Pity. otherwise I could have solved my problem with this program.

i can't load the nightscout collections from the vm to my pc.

As you already said, I usually ask in the corresponding Nightscoutforum

<img width="450" alt="image" src="https://user-images.githubusercontent.com/122942359/219124301-37fb6b10-8ede-4db9-84c2-32c01bb212eb.png">

**Reply by Navid200** [reply]
Maybe you can follow Patrick's suggestion and temporarily setup a different platform that uses Mongo on Atlas.  Transfer your data from Clarity to it.
Then, you can use our clone utility to copy data from that Nightsocut to your Google Cloud Nightscout.

If that works, you can consider it a 2-step solution.

**Reply by DaNitroo** [reply]
I have now tried this 2 step solution and created a second nightscout with railway just like patrick said.

The instructions were great and everything worked. I can also export and import csv, but unfortunately the data are too different to bring them into a common format. So I give up at this point and thank you for your help :-)

I need some sleep again, so maybe I'll reconcile the fact that the first few weeks that my son had diabetes are missing from the records. I have them at Dexcom and once we've had diabetes for 20 years, I'm not interested in the first few weeks either ;-)

**Reply by Navid200** [reply]
If your data has been transferred to Railway, you are done.
Now, all you need to do is to use this utility:  https://navid200.github.io/xDrip/docs/Nightscout/NS_Transfer.html

It asks you to provide a link for the site.  Provide the link to your railway site.  It will then read all the data from it.

**Reply by DaNitroo** [reply]
Unfortunately, that doesn't help me because I only have the data in Dexcom Clarity for the first few weeks of our diabetes. I can't get the dexcom csv in the format of MongoDB.csv, because unfortunately they both use very different data sets.

I will delete the railway again and use the Nightscout on the Google Cloud VM happily for the next few years, I think

**Reply by DaNitroo** [reply]
felix has diabetes since 14.11.2022. Since that day we have the Dexcom G7 and all data in Dexcom Clarity. I have been using XDrip & Watchdrip & nightscout since 12/15/22 and from then on I have all the data thanks to your instructions on google.

If I can't import for the first few weeks, then I just don't do it.

anyway thanks a lot for your help!!!

**Reply by Navid200** [reply]
If you have the data from Clarity to Railway, I told you how you can get that data into your Nightscout on Google Cloud Nightscout.
Doing that will just bring that data in without disturbing the existing data that you say is from 12/15/22.

I don't mean to drag this on.  I just am not sure we are communicating correctly with each other.  I don't understand what the problem is.
But, if you are happy, that's all I care about.

There will be other people who will find this thread.  If anyone else needs to transfer data from Clarity to Nightscout on Google Cloud, the option is there.

**Reply by DaNitroo** [reply]
The problem is that I can't get my Clarity.csv data to railway.

I can access the database and generally also import and export csv files. But unfortunately the Clarity.csv and the csv data that Nightscout saves are so different that the Clarity.csv cannot simply be inserted as I had imagined.

**Reply by Navid200** [reply]
OK I had misunderstood.  I thought you succeeded in getting the data to Railway.  Now I understand.
Thanks for clarifying.

**Reply by DaNitroo** [reply]
That will have been my fault. I always enter the text into the google translator and sometimes it doesn't come out what I wanted to say :-)

Thanks again anyway

**Comment by DaNitroo** [comment]
unfortunately they are totally different and have completely different definitions and a different structure

Dexcom:
<img width="902" alt="image" src="https://user-images.githubusercontent.com/122942359/219188593-3a989cf9-3fee-4a8f-b420-2911821d4d30.png">


NS:
<img width="721" alt="image" src="https://user-images.githubusercontent.com/122942359/219189147-93088302-cde9-4a14-a992-95aa597b755e.png">

**Comment by DaNitroo** [comment]
Now i still have a bit of hope to find a solution for my problem. But I don't know if I should ask this in the discussion here, or should I open a new discussion for it?


I found out that the data I'm missing in my nightscout in the Coogle Cloud is all in Xdrip on my son's phone. I was able to make a backup there and import it into xdrip and now have all the data since my son has in my xdrip.

Now the question is, how do I get all this CGm data to nightscout? Nightscout has only shown me the data since December 15th. but in Xdrip I see that we already had data 3 weeks before. I have already tried the backfill function of Xdrip. I get the message that the export was successful, but I still can't see the data in Nightscout.

I would be very happy about an idea.

**Reply by Navid200** [reply]
xDrip backfills the data to Nightscout that it is currently uploading to.
If your son's xDrip is uploading to Nightscout on Google Cloud and if you activated back-fill data under Settings &#8722;> Cloud Upload &#8722;> Nightscout Sync &#8722;> Extra Options, that will back-fill.

Check on the system status page to see how much of it has been completed.  After the system status page shows that there is nothing left in the queue, you should be able to see it in Nightscout.

**Reply by DaNitroo** [reply]
Unfortunately that doesn't work :-(

The following always happens: I take my phone and import my son's xDrip Googledrive backup. (He sleeps and has his mobile phone as a transmitter in his room). The data all appear correctly in my xDrip. Then I activate under Xdrip->Settings->Cloud Upload->API Upload-> Then I set REST-API to ON at the top and enter the base URL. I then go to "Backfill dates", enter the date and the message "Please wait" appears. after a few seconds the message appears: "Queued 23738 glucose readings and 2 treadments!" Then the xDrip start page appears and all glucose values ​​are gone and my xDrip is completely empty. Measurements are no longer displayed and the history is also completely empty.

<img width="157" alt="image" src="https://user-images.githubusercontent.com/122942359/219486913-e618f66c-831f-40f7-82ce-1ae713ad3fac.png">

**Reply by Navid200** [reply]
I don't understand why you want to upload to Nightscout from two different phones.
Why can you not just send a back-fill request from your son's xDrip?

**Reply by DaNitroo** [reply]
I just wanted to grab my phone so his could be left next to him while he slept.

Well I got it and tried what you said.

It really seems to work and the system status shows me that there are still 30230 files to be exported to Nightscout :-)

I'm curious, it should be ready in a few minutes :-)

**Reply by DaNitroo** [reply]
I can hardly believe it but it worked! :-)

Xdrip used this method to import all the data from my son's cell phone to Nightscout and then I activated the new Google Cloud backup in xDrip on both cell phones. Now the data is hopefully safe forever :-)

That actually makes an additional Nightscout backup superfluous, or am I missing something?

Thank you again!!!

**Reply by Navid200** [reply]
I'm glad you are happy.

I don't follow.  How can I use your method to automatically keep my repository protected against accidental deletion of the virtual machine?
My data goes back over 4 years.  I don't have all of that in my current phone.  But, I have data in Nightscout from different phones I have carried over the past 4 years.

**Reply by DaNitroo** [reply]
Luckily I had all the data in xDrip in one cell phone in my son's cell phone. With the function in xDrip, with which you can now syncronize xDrip with Google Drive, I can hopefully never lose data again. To be on the safe side, I save them as you showed me by downloading the Nightscout backup file - better safe than sorry :-)
